File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/person/BirthDateAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import java.time.LocalDate;
import java.util.List;

import com.truveta.opentoken.attributes.general.DateAttribute;
import com.truveta.opentoken.attributes.validation.DateRangeValidator;

/**
 * Represents the birth date attribute.
 * 
 * This class extends DateAttribute and provides functionality for working with
 * birth date fields. It recognizes "BirthDate" and "DateOfBirth" as valid aliases 
 * for this attribute type.
 * 
 * The attribute performs normalization on input values, converting them to a
 * standard format (yyyy-MM-dd) (inherited from DateAttribute).
 * 
 * The attribute also performs validation on input values, ensuring they:
 * - Match one of the supported date formats (inherited from DateAttribute)
 * - Fall within an acceptable range for birth dates (1910-01-01 to today)
 */
public class BirthDateAttribute extends DateAttribute {

    private static final String NAME = "BirthDate";
    private static final String[] ALIASES = new String[] { NAME, "DateOfBirth" };

    public BirthDateAttribute() {
        super(List.of(new DateRangeValidator(LocalDate.of(1910, 1, 1), true)));
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/person/birth_date_attribute.py
# src/opentoken/attributes/person/birth_date_attribute.py
from typing import List
from datetime import date
from opentoken.attributes.general.date_attribute import DateAttribute
from opentoken.attributes.validation.date_range_validator import DateRangeValidator


class BirthDateAttribute(DateAttribute):
    """Represents the birth date attribute.

    This class extends DateAttribute and provides functionality for working with
    birth date fields. It recognizes "BirthDate" as a valid alias for this
    attribute type.

    The attribute performs normalization on input values, converting them to a
    standard format (yyyy-MM-dd).

    Supported formats:
    - yyyy-MM-dd
    - yyyy/MM/dd
    - MM/dd/yyyy
    - MM-dd-yyyy
    - dd.MM.yyyy
    """

    NAME = "BirthDate"
    ALIASES = [NAME]

    def __init__(self):
        # Birth dates must be between 1910-01-01 and today
        min_date = date(1910, 1, 1)
        validator = DateRangeValidator(min_date=min_date, use_current_as_max=True)
        super().__init__(additional_validators=[validator])

    def get_name(self) -> str:
        return self.NAME

    def get_aliases(self) -> List[str]:
        return self.ALIASES.copy()
